test/functional/mempool_expiry.py

```diff
@@ -26,73 +26,78 @@
 class MempoolExpiryTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 1
+        self.setup_clean_chain = True
 
-    def skip_test_if_missing_module(self):
-        self.skip_if_no_wallet()
+    def trigger_mempool_expiry(self):
+        """Broadcasts an unrelated transaction as the expiry of the transactions in the mempool is only checked
+        when a new transaction is added to the mempool."""
+        try:
+            self.wallet.send_self_transfer(from_node=self.nodes[0])
+        except JSONRPCException:
```

Comment:
can you explain why this is needed? I'd prefer not to call the method when it won't do anything anyway

Reply:
[`CTxMemPool::Expire(..)`](https://github.com/bitcoin/bitcoin/blob/cd720337fe9c27422fe8d507726ca8c79f52b4ba/src/txmempool.cpp#L945) is only called in [ `LimitMempoolSize(..)`](https://github.com/bitcoin/bitcoin/blob/f17e8ba3a17b6516a1b1fb7f45d506a339e99f90/src/validation.cpp#L331) which is only called in [`UpdateMempoolForReorg(..)`](https://github.com/bitcoin/bitcoin/blob/f17e8ba3a17b6516a1b1fb7f45d506a339e99f90/src/validation.cpp#L370) and [`MemPoolAccept::Finalize(..)`](https://github.com/bitcoin/bitcoin/blob/f17e8ba3a17b6516a1b1fb7f45d506a339e99f90/src/validation.cpp#L983). 

To trigger the expiry of a transaction in the mempool we send an unrelated transaction. For the unrelated transaction `MemPoolAccept::Finalize(..)` is called which calls `LimitMempoolSize(..)` which calls `CTxMemPool::Expire(..)` on all timed-out transactions. If we don't send a new transaction, we can't test a transaction is really removed.

Reply:
I meant the `except: pass` part, sorry for being unclear :sweat_smile:

Reply:
There's a minor problem here, re the **unrelated transaction** part of @0xB10C's explanation. With `MiniWallet` if we don't do any extra bookkeeping in this test, and just do `self.wallet.send_self_transfer(from_node=node)` to trigger the mempool update, the (now expired) utxo will be spent. So this `send_self_transfer` will throw `test_framework.authproxy.JSONRPCException: mempool full (-26)`

We'd need to ensure that we don't spend one of the expired UTXOs when we want to `trigger_mempool_expiry` if we don't catch the Exception. ie:
```C++
node.setmocktime(expiry_time)
# Broadcasts an unrelated transaction as the expiry of the transactions in the mempool is only checked when a new transaction is added to the mempool.
new_utxo = self.wallet.get_utxo(txid=independent_txid)
new_txid = self.wallet.send_self_transfer(from_node=self.nodes[0], utxo_to_spend=new_utxo)['txid']
# parent_txid has now been evicted from the mempool.
```

Because the mempool is still updated if `send_self_transfer` fails, I thought it was cleaner to catch the Exception rather than have extra logic (that's unrelated to what we're testing) to ensure we're not spending expired UTXOs in this trigger.

Maybe I need to add a comment along these lines in the `except: pass` to make this clear? I'm realizing this is more confusing than I thought

Reply:
I don't understand why spending an non-existent coin will throw 'mempool full' (not missing-inputs) and how that is going to trigger mempool expiry of other txs.

Anyway, I guess the options are:

* Add a comment
* Remember the previous utxo (not expired) and use that
* Drop utxos that are about to expire from the wallet (might require fresh coins later on in the script)
* Something else?

Reply:
So I figured out why there were some weird errors that didn't make sense as you point out. After the first subtest some UTXOs were hanging around in the mempool, that ended up being used in the second subtest because MiniWallet sorts for the largest UTXOs. That's why I ensure the mempool is clear at the beginning of each subtest run now.